Post: protecting your frontend from a bad backend (#127)
xmartlabs/blog: _posts/2020-04-23-protecting-your-frontend-from-a-bad-backend.md

---
layout: post
title: Protecting your frontend from a bad backend
date: 2020-04-23 09:00:00
author: Matías Lorenzo
tags: [Xmartlabs, Architectural Patterns, Frontend]
author_id: mlorenzo
featured_image: /images/url-splitting/banner.jpg
show: true
category: development
---

> As a developer, you'll probably have to consume services provided by code of other developers.

Author: How about: 

```suggestion
As a developer, you'll probably have to consume services provided by third-party code.
```

> As a developer, you'll probably have to consume services provided by code of other developers.

Reviewer: I like it, and looking back at my suggestion I believe adding "from" would be incorrect

1. Query the API as it expects (the correct endpoint, HTTP method, etc).
2. Get data back and deserialize it (via a serializer).
3. Instantiate models from the data (could be only one or a collection of them).

We'll explain serializers and models later, but let's assume we have a `ProductSerializer` and a `Product` (serializer and model respectively).
Here's very simple code to implement our current ProductsController.

```js
class ProductsController {
  static async fetchProducts() {
    const response = await makeRequest('/products');
    // response.data is expected to be an array of product information.
    const deSerializedData = response.data.map(ProductSerializer.deSerialize);
    return deSerializedData.map(params => new Product(params));
  }
}
```

I've defined the controller as a class with a static method because I feel it's easier to read and understand, but you can certainly implement this however you feel it's best.
Let's assume the `makeRequest` function is defined elsewhere and performs an HTTP request to our API.
It's pretty clear what's going on, I think.
First, the API is queried.
Then, once the data arrives, it is deSerialized by our `ProductSerializer`.

> That resulting information is then fed into the `Product` model to create instances.

Author: Didn't think it would warrant one. I didn't want to use too much space to summarize what a controller does.

      description: data.Product_Desc,
      hasStock: data.Product_Has_Stock === 'true',
    };
  }
}
```

Our serializer has taken a single product and returned a new object with only the information we need, using a key nomenclature that is easier to read, and has also changed the type of the `hasStock` field to boolean.
Notice how the database timestamps are nowhere to be seen, we don't need them so we just ignore them.
Serializers are great since they automatically document the data we're using from our API.
No more "do we even use this field?" questions anymore, just go to the appropriate serializer and look it up.

What if, in a few months, the API decides that `Product_Name` should be changed to `Product_name` (lower case "n").
If we've been using this field everywhere in our app, then we might be at risk of introducing bugs by mass-replacing (this is quite a simple example, I know, but bear with me).
With serializers, this is a trivial issue.
Just go to the serializer and change how the name is obtained: `name: data.Product_name` and *voilà*.
Commit, open a PR and you're done for the day.

Also note how serializers are also resilient agains changes in structure.

> APIs can change where the data is stored in the response tree, but serializers can just go and fetch it from where they need.

Author: You're right, it's confusing. I'll change it :)

> APIs can change where the data is stored in the response tree, but serializers can just go and fetch it from where they need.

Author: Done!

  isAvailable() {
    // Complex logic here to determine if the product is available
  }
}
```

Having a model that represents our data allows us to encapsulate business logic related to the data right along the data itself.
Also, it allows us to store calculated attributes (for example, data that is derived from the initial data but that we pre-compute to avoid unnecessary computations later).
You might have noticed the constructor is quite silly.
We're just storing the parameters that were passed.
And if you look closely you'll find out the parameters have the same structure as what the serializer creates when deserializing the data.
This might seem repetitive and kinda pointless but I think it's really worth it, and the cost of implementing a model is almost zero.

One of the greatest things about models is that you no longer work with plain Javscript objects. Your data has meaning and structure now. Data logging is now a bit more clear, as I can instantly know what type of data I'm logging. If some data I'm handling doesn't have a model prototype, then I know something went wrong somewhere.

### Putting all together

Let's summarize what each part of the pattern provides:

> * **Controllers**: provide an abstraction of our API (or a portion of it) and allow other components to forget where the data comes from and how it is fetched.

Reviewer: ```suggestion
* **Controllers**: provide an abstraction of our API (or a portion of it), which allows other components to not care about where the data comes from and how it is fetched.
```

> * **Serializers**: protect data by acting as an adapter and a firewall. They transform and remove data according to our needs and document what we're using.

Reviewer: ```suggestion
* **Serializers**: protect data by acting as an adapter and a firewall. They transform and remove data according to our needs, while at the same time documenting what we're using.
```

* **Models**: add meaning to our data. They aid us by allowing the frontend to inject its own logic and value to data, without needing support from the API.

Remember that the idea of the pattern is to be versatile and applicable in many different cases.
I've shown a possible minimal implementation with Javascript, but you can certainly use it in other languages with any framework.
As long as you have *something* in your code that acts as a part of the pattern, then you're implementing it!
I recommend (if you want to apply it) to experiment and see what suits you best.

> In my experience in my team I've seen that people tend to make this pattern their own by implementing it slightly differently, which is awesome and helps us learn from others.

Reviewer: "in my team" sounds redundant
```suggestion
In my experience, I've seen that people tend to make this pattern their own by implementing it slightly differently, which is awesome and helps us learn from others.
```

### Error handling

We haven't discussed yet how to handle errors or unhappy paths in our app.

> I'm not going to go into full detail as this post is already quite packed, so I'm just going to quickly explain how we handle errors in our projects with this pattern.

Reviewer: ```suggestion
I'm not going to go into full detail, as this post is already quite packed. So I'm just going to quickly explain how we handle errors in our projects with this pattern.
```

Typically we implement some kind of networking service that acts as an adapter with the networking library we're using (fetch, axios, whatever you like).
In the controller we discussed above, this service would be the implementation of the `makeRequest` method.
When any kind of error happens, we catch it and attempt to deserialize it (this typically happens when the error has status 4xx, meaning it was a client error) to understand what happened.
The deserialized error is then thrown again to be catched by the entity that called on the controller.
That way the calling entity doesn't have to know what underlying networking library we're using, since we intercept those errors, extract relevant information and create a new error.
You could even switch networking libraries without changing anything but the networking service!

Error handling is sometimes hard, and it requires support from the backend.
We must know how errors are categorized and come up with a certain "contract" for how the backend communicates these errors.

> This is sadly unavoidable, but using this pattern you can also protect yourself from bad errors by deserializing them.

Reviewer: ```suggestion
This is sadly unavoidable, but using this pattern you can also protect yourself from badly formatted errors by deserializing them.
```
